### ISSUE TITLE
Serve Name Indication

### ISSUE THREAD

**Issue by Costelo** [sqlmapproject/sqlmap#1256]
`````
python sqlmap.py -u "https://bake-house.com/" --random-agent -v 3   

[15:38:18] [DEBUG] cleaning up configuration parameters
[15:38:18] [DEBUG] checking for WebSocket
[15:38:18] [DEBUG] setting the HTTP timeout
[15:38:18] [DEBUG] loading random HTTP User-Agent header(s) from file '/tmp/sqlmap/txt/user-agents.txt'
[15:38:18] [INFO] fetched random HTTP User-Agent header from file '/tmp/sqlmap/txt/user-agents.txt': 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US) AppleWebKit/530.5 (KHTML, like Gecko) Chrome/2.0.172.42 Safari/530.5'
[15:38:18] [DEBUG] creating HTTP requests opener object
[15:38:18] [DEBUG] resolving hostname 'bake-house.com'
[15:38:19] [INFO] testing connection to the target URL
[15:38:20] [DEBUG] SSL connection error occurred ('[SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:590)')
[15:38:20] [DEBUG] SSL connection error occurred ('[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:590)')
[15:38:21] [DEBUG] SSL connection error occurred ('[SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:590)')
[15:38:21] [CRITICAL] can't establish SSL connection
``````

the problem is with SNI (Server Name Indication). Other web-sites, example https://google.com ok, but those that use CloudFlare, not working.

Python 2.7.10, SQLMap last version, OpenSSL 1.0.2a

**Comment by stamparm** [comment]
To quote http://bugs.python.org/issue5639:

`Client-side SNI support has been added in Python 3.2, and 3.4 is now out. People who migrated their code to Python 3 have been enjoying SNI support for years now, and they're gradually getting more TLS features at every new feature release.`

p.s. this doesn't mean that if there is a way-around that we shouldn't implement it into the sqlmap

**Comment by stamparm** [comment]
It should be implemented now. Only requirement is to run Python >= v2.7.9 (Reference: https://docs.python.org/2/library/ssl.html#ssl-contexts)

**Comment by Mrten** [comment]
@stamparm: With this patch I get (and only get!) 403 errors on a server that has SNI enabled. Only if I comment out the complete ` +        if not kb.tlsSNI: ` if I manage to get a working connection to the servers I want to test. Which is not weird, because a 403 is not an SSL exception...

**Comment by stamparm** [comment]
@Mrten well, you can comment out that line, but SNI (as far as I know) require a hostname which those lines that you include when commenting out the `if not kb.tlsSNI` do not provide.

So, when you say that you `manage to get a working connection to the servers`, I am afraid that you are not getting the connection to the right server.

Please contact me to `miroslav@sqlmap.org` for further discussion (and potential problem solving)

**Comment by Mrten** [comment]
I'm sorry for being unclear: I meant that I commented out everything between the line I mentioned and the ` +        if not success and hasattr(ssl, "SSLContext"): ` line.

**Comment by stamparm** [comment]
@Mrten but those lines should not case problems. Those are required for non-TLS (and older versions of Python).

Only case where I can spot the problem is that your server accepts non-SNI connection

**Comment by stamparm** [comment]
@Mrten I'll make one small patch and let you know

**Comment by stamparm** [comment]
@Mrten can you please update and retry? Now the SNI is preferred (if available)

**Comment by Mrten** [comment]
Using 1.0-dev-c7ec153: much better :)

**Comment by Mrten** [comment]
Btw, I'm using python 2.7.9 on debian jessie.

**Comment by Mrten** [comment]
Want me to try 2.7.3?

**Comment by stamparm** [comment]
@Mrten I believe that you'll have problems as SNI support is not available on <2.7.9

**Comment by Mrten** [comment]
Yeah, that's why I asked... You mentioned 'older python versions' :)
